addons/l10n_es_edi_sii/models/account_edi_format.py

```diff
@@ -0,0 +1,587 @@
+            client = zeep.Client(connection_vals['url'], transport=transport)
+
+            if invoices[0].is_sale_document():
+                service_name = 'SuministroFactEmitidas'
+            else:
+                service_name = 'SuministroFactRecibidas'
+            if company.l10n_es_edi_test_env and not connection_vals.get('test_url'):
+                service_name += 'Pruebas'
+
+            # Establish the connection.
+            serv = client.bind('siiService', service_name)
+            if company.l10n_es_edi_test_env and connection_vals.get('test_url'):
+                serv._binding_options['address'] = connection_vals['test_url']
+
+            try:
+                if invoices[0].is_sale_document():
+                    res = serv.SuministroLRFacturasEmitidas(header, info_list)
+                else:
+                    res = serv.SuministroLRFacturasRecibidas(header, info_list)
+            except Exception as error:
+                if isinstance(error, requests.exceptions.SSLError):
+                    msg = _("The SSL certificate could not be validated.")
+                elif isinstance(error, zeep.exceptions.Error):
+                    msg = _("Networking error:\n%s") % error
+                else:
+                    msg = str(error)
+                return {inv: {
+                    'error': msg,
+                    'blocking_level': 'warning',
+                } for inv in invoices}
+
+            # Process response.
+
+            if not res or not res.RespuestaLinea:
+                return {inv: {
+                    'error': _("The web service is not responding"),
+                    'blocking_level': 'warning',
+                } for inv in invoices}
+
+            resp_state = res["EstadoEnvio"]
+            l10n_es_edi_csv = res['CSV']
+            for respl in res.RespuestaLinea:
+                invoice_number = respl['IDFactura']['NumSerieFacturaEmisor']
+
+                # Retrieve the corresponding invoice.
+                inv = invoices.filtered(lambda x: (x.name if x.is_sale_document else x.ref) == invoice_number)
+
+                if resp_state == 'Correcto':
+                    inv.l10n_es_edi_csv = l10n_es_edi_csv
+                    return {inv: {'success': True} for inv in invoices}
+                elif resp_state == 'ParcialmenteCorrecto' and respl['EstadoRegistro'] == 'AceptadoConErrores':
+                    inv.l10n_es_edi_csv = l10n_es_edi_csv
+                    for inv in invoices:
+                        inv.message_post(body=_("This was accepted with errors."))
+                    return {inv: {'success': True} for inv in invoices}
+                elif resp_state == 'Incorrecto':
+                    return {inv: {
+                        'error': _("[%s] %s", respl['CodigoErrorRegistro'], respl["DescripcionErrorRegistro"]),
+                        'blocking_level': 'warning',
+                    } for inv in invoices}
```

Comment:
That is a really weird loop: we're iterating on the lines, but if the toplevel response state is in any relevant one we're going to update the invoice(s) of the first line then immediately return. So we're only ever processing the invoice of the first line, before returning a result signal for the entire batch of invoices.

Is that normal? Why are we iterating at all on the list of lines in that case? Just get the first line directly.

Also why is there a mix of accessing to attributes and statically named items? e.g. why `res.RespuestaLinea` but `res["EstadoEnvio"]`? Are the semantics different?

Reply:
@xmo-odoo I need to improve this.  You can send multiple invoices and one can be correct and the other not.

Reply:
Right so this loop probably needs to build up the result on a per-invoice basis, possibly with fast paths for the case where we know up-front everything is right or wrong? How is the "partial" version signaled?

Reply:
If it is not all correct, then either you have partially correct or incorrect, but in those cases I go through the states on the line now anyways.  It is because we handle it in json that there is the confusion of [" or ., but indeed, we could use . throughout